src/sacn/source.c

```diff
@@ -595,13 +1232,695 @@ etcpal_error_t sacn_source_reset_networking(sacn_source_t handle, uint16_t unive
  * @return The total number of network interfaces for the universe. If this is greater than netints_size, then only
  * netints_size addresses were written to the netints array. If the source or universe were not found, 0 is returned.
  */
-size_t sacn_source_get_network_interfaces(sacn_source_t handle, uint16_t universe, SacnMcastInterface* netints,
+size_t sacn_source_get_network_interfaces(sacn_source_t handle, uint16_t universe, EtcPalMcastNetintId* netints,
                                           size_t netints_size)
 {
-  ETCPAL_UNUSED_ARG(handle);
-  ETCPAL_UNUSED_ARG(universe);
-  ETCPAL_UNUSED_ARG(netints);
-  ETCPAL_UNUSED_ARG(netints_size);
+  size_t total_num_network_interfaces = 0;
+
+#if SACN_SOURCE_ENABLED
+  if (sacn_lock())
+  {
+    // Look up universe state
+    SacnSourceUniverse* universe_state = NULL;
+    if (lookup_source_state(handle, universe, NULL, &universe_state) == kEtcPalErrOk)
+    {
+      total_num_network_interfaces = universe_state->num_netints;
+
+      // Copy out the netints
+      for (size_t i = 0; netints && (i < netints_size) && (i < total_num_network_interfaces); ++i)
+        netints[i] = universe_state->netints[i];
+    }
+
+    sacn_unlock();
+  }
+#endif
+
+  return total_num_network_interfaces;
+}
+
+bool source_handle_in_use(int handle_val, void* cookie)
+{
+  ETCPAL_UNUSED_ARG(cookie);
+
+  return (handle_val == SACN_SOURCE_INVALID) || (lookup_source_state(handle_val, 0, NULL, NULL) == kEtcPalErrOk);
+}
+
+// Needs lock
+etcpal_error_t start_tick_thread()
+{
+  shutting_down = false;
+  EtcPalThreadParams params = ETCPAL_THREAD_PARAMS_INIT;
+
+  if (etcpal_thread_create(&source_thread_handle, &params, source_thread_function, NULL) != kEtcPalErrOk)
+    return kEtcPalErrSys;
+
+  return kEtcPalErrOk;
+}
+
+// Takes lock
+void stop_tick_thread()
+{
+  etcpal_thread_t thread_handle;
+
+  if (sacn_lock())
+  {
+    shutting_down = true;  // Trigger thread-based sources to terminate
+    thread_handle = source_thread_handle;
+    sacn_unlock();
+  }
+
+  // Wait for thread-based sources to terminate (assuming application already cleaned up manual sources)
+  etcpal_thread_join(&thread_handle);
+}
+
+// Takes lock
+void source_thread_function(void* arg)
+{
+  ETCPAL_UNUSED_ARG(arg);
+
+  bool keep_running_thread = true;
+  int num_thread_based_sources = 0;
+
+  EtcPalTimer interval_timer;
+  etcpal_timer_start(&interval_timer, SOURCE_THREAD_INTERVAL);
+
+  while (keep_running_thread || (num_thread_based_sources > 0))
```

Comment:
Hmm, so the thread stops if there are no more thread-based sources, but what starts it again if another thread-based source is added? This logic feels like an overly complex optimization, I think we should start the thread when it is first requested and keep running it until the library is deinitialized.

Reply:
If shutting_down is false then the thread will keep going even if there are no more thread-based sources (see keep_running_thread). The only way to get the thread to stop is to enable shutting_down, which is only set to true from sacn_source_deinit. That will automatically trigger termination for all thread-based sources as well, so at that point the thread keeps running until num_thread_based_sources inevitably falls to 0. So I think it already does what you want it to do.

Reply:
Oh I see. The logic is so that you make sure you get the required Stream_Terminated packets out on shutdown. Gotcha

Reply:
Improved documentation for this in https://github.com/ETCLabs/sACN/pull/32/commits/c634b9c92e56fdedecc51860acf8f997808cc002.